libraries/cmake/source/lief/CMakeLists.txt

```diff
@@ -0,0 +1,388 @@
+  CONFIGURE_COMMAND ""
+  BUILD_COMMAND     ""
+  UPDATE_COMMAND    ""
+  INSTALL_COMMAND   "")
+
+ExternalProject_get_property(lief_leaf SOURCE_DIR)
+set(LEAF_INCLUDE_DIR "${SOURCE_DIR}/include")
+
+set(UTFCPP_VERSION 3.1.2) # Custom fix to remove use of SUBLANG_DEFAULT in common.hpp and all.hpp
+set(UTFCPP_SHA256 SHA256=b77bff122a6d4f2a7a1ab409086bbb59bf899a2fdde12e1a85a4305fa91764c4)
+set(UTFCPP_URL "${library_root}/third-party/utfcpp-${UTFCPP_VERSION}.zip" CACHE STRING "URL to UTFCPP")
+ExternalProject_Add(lief_utfcpp
+  URL               ${UTFCPP_URL}
+  URL_HASH          ${UTFCPP_SHA256}
+  CONFIGURE_COMMAND ""
+  BUILD_COMMAND     ""
+  UPDATE_COMMAND    ""
+  INSTALL_COMMAND   "")
+
+ExternalProject_get_property(lief_utfcpp SOURCE_DIR)
+set(UTFCPP_INCLUDE_DIR "${SOURCE_DIR}/source")
+
+add_dependencies(thirdparty_lief lief_mbed_tls lief_spdlog lief_leaf lief_utfcpp)
+target_link_libraries(thirdparty_lief PRIVATE
+  thirdparty_cxx_settings
+  thirdparty_c_settings
+  )
+target_link_libraries(thirdparty_lief PUBLIC
+  thirdparty_mbedtls
+  )
+
+add_custom_command(OUTPUT ${library_root}/include/LIEF/third-party/boost/leaf/all.hpp
+  COMMAND
+    ${CMAKE_COMMAND} -E copy_directory ${LEAF_INCLUDE_DIR}/
+    ${library_root}/include/LIEF/third-party/
+  DEPENDS lief_leaf)
+
+target_sources(thirdparty_lief PRIVATE
+  ${library_root}/include/LIEF/third-party/boost/leaf/all.hpp)
+
+add_custom_command(OUTPUT ${library_root}/include/LIEF/third-party/utfcpp/utf8.h
+  COMMAND
+    ${CMAKE_COMMAND} -E copy_directory ${UTFCPP_INCLUDE_DIR}/
+    ${library_root}/include/LIEF/third-party/utfcpp
+  DEPENDS lief_utfcpp)
+
+target_sources(thirdparty_lief PRIVATE
+  ${library_root}/include/LIEF/third-party/utfcpp/utf8.h)
+
+target_include_directories(thirdparty_lief PRIVATE
+  "${MBEDTLS_INCLUDE_DIRS}"
+  "${SPDLOG_INCLUDE_DIR}"
+  "${LEAF_INCLUDE_DIR}"
+  "${UTFCPP_INCLUDE_DIR}"
+  )
+
+set(LIEF_SUPPORT_CXX11 0)
+set(LIEF_SUPPORT_CXX14 0)
+set(LIEF_SUPPORT_CXX17 0)
+
+if (cxx_std_11 IN_LIST CMAKE_CXX_COMPILE_FEATURES)
+  set(LIEF_SUPPORT_CXX11 1)
+endif()
+
+if (cxx_std_14 IN_LIST CMAKE_CXX_COMPILE_FEATURES)
+  if (${MSVC} AND ${MSVC_TOOLSET_VERSION} GREATER_EQUAL 141)
+    set(LIEF_SUPPORT_CXX14 1)
+  elseif((NOT DEFINED MSVC) OR (NOT ${MSVC}))
+    set(LIEF_SUPPORT_CXX14 1)
+  endif()
+endif()
+
+if (cxx_std_17 IN_LIST CMAKE_CXX_COMPILE_FEATURES)
+  if (${MSVC} AND ${MSVC_TOOLSET_VERSION} GREATER_EQUAL 142)
+    set(LIEF_SUPPORT_CXX17 1)
+  elseif((NOT DEFINED MSVC) OR (NOT ${MSVC}))
+    set(LIEF_SUPPORT_CXX17 1)
+  endif()
+endif()
+
+configure_file(
+  "${library_root}/src/compiler_support.h.in"
+  "${library_root}/src/compiler_support.h"
+  @ONLY
+  )
+
+set_target_properties(
+  thirdparty_lief
+  PROPERTIES POSITION_INDEPENDENT_CODE ON
+             CXX_STANDARD              11
+             CXX_STANDARD_REQUIRED     ON
+             CXX_VISIBILITY_PRESET     hidden
+             C_VISIBILITY_PRESET       hidden)
+endfunction()
+
+liefCppMain()
```

Comment:
I'm doing a general comment here for the line from 243 and on.

In general the theme here is that we should not do any generation of code, or discovery of flags during configure time, but just "snapshot" what was the state of the project (compiler flags, generated files and so on)  when it has been configured the first time on the target platform, before including it with osquery.

Also, all these `ExternalProject_Add` are downloading additional parts/libraries of the project. These, if necessary, should all become submodules and added as libraries as you're doing with LIEF.

Reply:
removed all ExternalProjects. I looked at the dependencies again, it looks leaf and utfcpp are header only dependencies. I just ended up placing them in `<root>/libraries/cmake/source/lief/config/lief/thirdparty`. Is that ok?
Or should they still be submodules?
For spdlog and mbedtls i made submodules.